File: app/src/main/java/com/xuie/androiddemo/util/ScreenUtils.java
package com.xuie.androiddemo.util;

import android.app.Activity;
import android.content.Context;
import android.content.res.Resources;
import android.graphics.Point;
import android.util.TypedValue;
import android.view.Display;

import java.lang.reflect.Field;

/**
 * Created by xuie on 15-11-4.
 */
public class ScreenUtils {

    public static Point getPoint(Activity activity) {
        Display display = activity.getWindowManager().getDefaultDisplay();
        Point size = new Point();
        display.getSize(size);
        return size;
    }

    public static int getWidth(Activity activity) {
        return getPoint(activity).x;
    }

    public static int getHeight(Activity activity) {
        return getPoint(activity).y;
    }

    public static int getStatusBarHeight(Context context) {
        Class<?> c = null;
        Object obj = null;
        Field field = null;
        int x = 0, statusBarHeight = 0;
        try {
            c = Class.forName("com.android.internal.R$dimen");
            obj = c.newInstance();
            field = c.getField("status_bar_height");
            x = Integer.parseInt(field.get(obj).toString());
            statusBarHeight = context.getResources().getDimensionPixelSize(x);
        } catch (Exception e1) {
            e1.printStackTrace();
        }
        return statusBarHeight;
    }

    public static int getNavigationBarHeight(Activity activity) {
        Resources resources = activity.getResources();
        int resourceId = resources.getIdentifier("navigation_bar_height", "dimen", "android");
        //获取NavigationBar的高度
        return resources.getDimensionPixelSize(resourceId);
    }

    public static float dpToPx(float dp) {
        return dp * Resources.getSystem().getDisplayMetrics().density;
    }

    public static float pxToDp(float px) {
        return px / Resources.getSystem().getDisplayMetrics().density;
    }

    public static float dpToPxInt(float dp) {
        return (int) (dpToPx(dp) + 0.5f);
    }

    public static float pxToDpCeilInt(float px) {
        return (int) (pxToDp(px) + 0.5f);
    }

    public static int getPixelFromDPI(float dpi) {
        return (int) (TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dpi, Resources.getSystem().getDisplayMetrics()));
    }
}

File: app/src/main/java/com/xuie/android/util/ScreenUtils.kt
package com.xuie.android.util

import android.app.Activity
import android.content.Context
import android.content.res.Resources
import android.graphics.Point
import android.util.TypedValue
import android.view.Display

import java.lang.reflect.Field

/**
 * @author yf
 * @date 15-11-4
 */
object ScreenUtils {

  fun getPoint(activity: Activity): Point {
    val display = activity.windowManager.defaultDisplay
    val size = Point()
    display.getSize(size)
    return size
  }

  fun getWidth(activity: Activity): Int {
    return getPoint(activity).x
  }

  fun getHeight(activity: Activity): Int {
    return getPoint(activity).y
  }

  fun getStatusBarHeight(context: Context): Int {
    var c: Class<*>? = null
    var obj: Any? = null
    var field: Field? = null
    var x = 0
    var statusBarHeight = 0
    try {
      c = Class.forName("com.android.internal.R\$dimen")
      obj = c!!.newInstance()
      field = c.getField("status_bar_height")
      x = Integer.parseInt(field!!.get(obj).toString())
      statusBarHeight = context.resources.getDimensionPixelSize(x)
    } catch (e1: Exception) {
      e1.printStackTrace()
    }

    return statusBarHeight
  }

  fun getNavigationBarHeight(activity: Activity): Int {
    val resources = activity.resources
    val resourceId = resources.getIdentifier("navigation_bar_height", "dimen", "android")
    //获取NavigationBar的高度
    return resources.getDimensionPixelSize(resourceId)
  }

  fun dpToPx(dp: Float): Float {
    return dp * Resources.getSystem().displayMetrics.density
  }

  fun pxToDp(px: Float): Float {
    return px / Resources.getSystem().displayMetrics.density
  }

  fun dpToPxInt(dp: Float): Float {
    return (dpToPx(dp) + 0.5f).toInt().toFloat()
  }

  fun pxToDpCeilInt(px: Float): Float {
    return (pxToDp(px) + 0.5f).toInt().toFloat()
  }

  fun getPixelFromDPI(dpi: Float): Int {
    return TypedValue.applyDimension(TypedValue.COMPLEX_UNIT_DIP, dpi, Resources.getSystem().displayMetrics).toInt()
  }
}
